File: src/utils_hardware.py
"""
ZEN_RAG Hardware Detection Module
==================================
Single source of truth for hardware detection across all modules.
Consolidates duplicated code from utils.py, api_server.py, and startup/.

UNIFIED FUNCTIONS:
    - get_hardware_profile()     : Cross-platform hardware detection
    - detect_cpu()               : CPU vendor/cores detection
    - detect_memory()            : RAM detection
    - detect_gpu()               : GPU detection (Windows with VRAM)
    - get_recommendation_tier()  : Model-size recommendation based on hardware

USAGE:
    from utils_hardware import get_hardware_profile
    hw = get_hardware_profile()
    print(f"Running on {hw['cpu_vendor']} with {hw['ram_gb']}GB RAM")
"""

import sys
import os
import platform
import subprocess
import json
from typing import Dict, Any, Optional
import logging

logger = logging.getLogger(__name__)

# ============================================================================
# CPU DETECTION
# ============================================================================


def detect_cpu() -> Dict[str, Any]:
    """Detect CPU vendor, cores, frequency.

    Returns:
        Dict with keys: vendor, name, logical_cores, physical_cores, freq_mhz
    """
    cpu_info = {
        "vendor": "Unknown",
        "name": platform.processor() or "Unknown",
        "logical_cores": os.cpu_count() or 1,
        "physical_cores": os.cpu_count() or 1,
        "freq_mhz": None,
    }

    try:
        import psutil

        # Try to get physical core count (psutil has fallback)
        physical = psutil.cpu_count(logical=False)
        if physical:
            cpu_info["physical_cores"] = physical

        # Try to get CPU frequency
        freq = psutil.cpu_freq()
        if freq:
            cpu_info["freq_mhz"] = round(freq.current, 0)
    except ImportError as exc:
        logger.debug("%s", exc)

    # Vendor detection from name
    name_upper = cpu_info["name"].upper()
    if "INTEL" in name_upper:
        cpu_info["vendor"] = "Intel"
    elif "AMD" in name_upper:
        cpu_info["vendor"] = "AMD"
    elif "ARM" in name_upper:
        cpu_info["vendor"] = "ARM"

    return cpu_info


# ============================================================================
# MEMORY DETECTION
# ============================================================================


def detect_memory() -> Dict[str, Any]:
    """Detect RAM availability and usage.

    Returns:
        Dict with keys: total_gb, available_gb, used_percent
    """
    mem_info = {
        "total_gb": 8.0,  # Safe default
        "available_gb": 4.0,
        "used_percent": 50.0,
    }

    try:
        import psutil

        vm = psutil.virtual_memory()
        mem_info["total_gb"] = round(vm.total / (1024**3), 1)
        mem_info["available_gb"] = round(vm.available / (1024**3), 1)
        mem_info["used_percent"] = vm.percent
    except ImportError:
        logger.warning("psutil not available for detailed memory detection")

    return mem_info


# ============================================================================
# GPU DETECTION (Windows)
# ============================================================================


def detect_gpu() -> Optional[Dict[str, Any]]:
    """Detect dedicated GPU (Windows only with PowerShell/WMI).

    Returns:
        Dict with keys: vendor, name, vram_mb
        Or None if no discrete GPU found
    """
    if sys.platform != "win32":
        return None

    try:
        cmd = [
            "powershell", "-NoProfile", "-Command",
            'Get-CimInstance Win32_VideoController | '
            'Where-Object {$_.Name -notmatch "Remote|RDP"} | '
            'Select-Object Name, AdapterRAM | ConvertTo-Json',
        ]

        output = subprocess.check_output(cmd, shell=False, text=True, stderr=subprocess.DEVNULL, timeout=5).strip()

        if not output:
            return None

        gpus = json.loads(output)
        if not isinstance(gpus, list):
            gpus = [gpus]

        # Filter out integrated graphics
        for gpu_raw in gpus:
            name = gpu_raw.get("Name", "").upper()

            # Skip Intel iGPU and Microsoft Basic Display Adapter
            if any(x in name for x in ["INTEL", "BASIC", "UHD", "IRIS"]):
                continue

            vram_bytes = gpu_raw.get("AdapterRAM", 0)
            vram_mb = max(1, vram_bytes // (1024**2))  # At least 1MB

            gpu_info = {
                "name": gpu_raw.get("Name", "Unknown"),
                "vram_mb": vram_mb,
                "vendor": "Unknown",
            }

            # Detect vendor
            if "NVIDIA" in name or "GEFORCE" in name or "TESLA" in name:
                gpu_info["vendor"] = "NVIDIA"
            elif "AMD" in name or "RADEON" in name or "EPYC" in name:
                gpu_info["vendor"] = "AMD"
            elif "INTEL" in name:
                gpu_info["vendor"] = "Intel"

            return gpu_info

    except Exception as e:
        logger.debug(f"GPU detection failed: {e}")
        return None


# ============================================================================
# RECOMMENDATION ENGINE
# ============================================================================


def get_recommendation_tier(ram_gb: float, cpu_cores: int) -> Dict[str, Any]:
    """Recommend LLM tier based on hardware.

    Args:
        ram_gb: Total system RAM in GB
        cpu_cores: Logical CPU cores

    Returns:
        Dict with: tier, max_model_params, quantization, n_gpu_layers, note
    """

    # Tier decision logic
    if ram_gb < 4:
        tier = "potato"
        max_params = "1B-3B"
        quant = "Q4_K_S"
        gpu_layers = 0
        note = "Very tight resources. Consider 1B models like TinyLlama"

    elif ram_gb < 8:
        tier = "budget"
        max_params = "3B-7B"
        quant = "Q4_K_S or Q5_K_M"
        gpu_layers = 0
        note = "Focus on 3B-7B quantized models"

    elif ram_gb < 16:
        tier = "mid_range"
        max_params = "7B-13B"
        quant = "Q5_K_M or Q6_K"
        gpu_layers = 10 if ram_gb > 12 else 5
        note = "Good range: 7B Mistral, Llama 2-7B"

    elif ram_gb < 32:
        tier = "comfortable"
        max_params = "13B-34B"
        quant = "Q5_K_M, Q6_K, or Q8_0"
        gpu_layers = 20 if ram_gb > 24 else 15
        note = "Can run larger models: Llama 2-13B, Mistral 7B without Q"

    else:
        tier = "rich"
        max_params = "34B-70B+"
        quant = "Q6_K, Q8_0, or full precision"
        gpu_layers = 40  # Offload most layers
        note = "Plenty of resources: Llama 2-70B, Code Llama, Mixtral"

    return {
        "tier": tier,
        "ram_gb": ram_gb,
        "cpu_cores": cpu_cores,
        "max_model_params": max_params,
        "recommended_quant": quant,
        "n_gpu_layers": gpu_layers,
        "note": note,
    }


# ============================================================================
# MAIN API
# ============================================================================


def get_hardware_profile(detailed: bool = False) -> Dict[str, Any]:
    """Get complete hardware profile for ZEN_RAG.

    Args:
        detailed: If True, include recommendation tier and extra info

    Returns:
        Comprehensive hardware dictionary:
        {
            "platform": "Windows/Linux/Darwin",
            "architecture": "x86_64/arm64/etc",
            "cpu": {...},
            "memory": {...},
            "gpu": {...} or None,
            "recommendation": {...} if detailed=True
        }
    """

    profile = {
        "platform": sys.platform,
        "architecture": platform.machine(),
        "cpu": detect_cpu(),
        "memory": detect_memory(),
        "gpu": detect_gpu(),
    }

    if detailed:
        profile["recommendation"] = get_recommendation_tier(
            profile["memory"]["total_gb"], profile["cpu"]["logical_cores"]
        )

    return profile


def get_hardware_summary() -> Dict[str, Any]:
    """Quick hardware summary for CLI tools (cached).

    Used by: api_server /health endpoint
    Returns minimal key info for status checks.
    """
    profile = get_hardware_profile()

    summary = {
        "ram_total_gb": profile["memory"]["total_gb"],
        "ram_available_gb": profile["memory"]["available_gb"],
        "cpu_count": profile["cpu"]["logical_cores"],
        "cpu_physical": profile["cpu"]["physical_cores"],
        "cpu_vendor": profile["cpu"]["vendor"],
    }

    if profile["gpu"]:
        summary["gpu_name"] = profile["gpu"]["name"]
        summary["gpu_vram_mb"] = profile["gpu"]["vram_mb"]

    return summary


# ============================================================================
# FOR BACKWARD COMPATIBILITY
# ============================================================================


class HardwareProfiler:
    """Backward-compatible wrapper for legacy utils.py code."""

    @staticmethod
    def get_profile() -> dict:
        """Legacy method - use get_hardware_profile() instead."""
        profile = get_hardware_profile(detailed=False)

        # Transform to legacy format
        return {
            "type": profile["cpu"]["vendor"],
            "ram_gb": profile["memory"]["total_gb"],
            "vram_mb": profile["gpu"]["vram_mb"] if profile["gpu"] else 0,
            "threads": profile["cpu"]["logical_cores"],
            "gpu": profile["gpu"]["name"] if profile["gpu"] else None,
        }


if __name__ == "__main__":
    # CLI test
    print("\n" + "=" * 70)
    print("ZEN_RAG HARDWARE PROFILE")
    print("=" * 70)

    hw = get_hardware_profile(detailed=True)

    print("\n🖥️  PLATFORM")
    print(f"   OS: {hw['platform']}")
    print(f"   Arch: {hw['architecture']}")
    print("\n💻 CPU")
    print(f"   {hw['cpu']['name']}")
    print(f"   Vendor: {hw['cpu']['vendor']}")
    print(f"   Logical Cores: {hw['cpu']['logical_cores']}")
    print(f"   Physical Cores: {hw['cpu']['physical_cores']}")
    if hw["cpu"]["freq_mhz"]:
        print(f"   Frequency: {hw['cpu']['freq_mhz']} MHz")
        pass
    print("\n🧠 MEMORY")
    print(f"   Total: {hw['memory']['total_gb']} GB")
    print(f"   Available: {hw['memory']['available_gb']} GB")
    print(f"   Used: {hw['memory']['used_percent']}%")
    if hw["gpu"]:
        print("\n🎮 GPU (Discrete)")
        print(f"   {hw['gpu']['name']}")
        print(f"   Vendor: {hw['gpu']['vendor']}")
        print(f"   VRAM: {hw['gpu']['vram_mb']} MB")
    else:
        print("\n🎮 GPU: None (CPU only)")

    if "recommendation" in hw:
        rec = hw["recommendation"]
        print("\n🎯 RECOMMENDATION")
        print(f"   Tier: {rec['tier'].upper()}")
        print(f"   Max Model: {rec['max_model_params']}")
        print(f"   Quantization: {rec['recommended_quant']}")
        print(f"   GPU Layers: {rec['n_gpu_layers']}")
        print(f"   Note: {rec['note']}")
    print("\n" + "=" * 70 + "\n")

File: rust_output/src/utils_hardware.rs
/// ZEN_RAG Hardware Detection Module
/// ==================================
/// Single source of truth for hardware detection across all modules.
/// Consolidates duplicated code from utils::py, api_server::py, and startup/.
/// 
/// UNIFIED FUNCTIONS:
/// - get_hardware_profile()     : Cross-platform hardware detection
/// - detect_cpu()               : CPU vendor/cores detection
/// - detect_memory()            : RAM detection
/// - detect_gpu()               : GPU detection (Windows with VRAM)
/// - get_recommendation_tier()  : Model-size recommendation based on hardware
/// 
/// USAGE:
/// from utils_hardware import get_hardware_profile
/// hw = get_hardware_profile()
/// print(f"Running on {hw['cpu_vendor']} with {hw['ram_gb']}GB RAM")

use anyhow::{Result, Context};
use std::collections::HashMap;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Backward-compatible wrapper for legacy utils::py code.
#[derive(Debug, Clone)]
pub struct HardwareProfiler {
}

impl HardwareProfiler {
    /// Legacy method - use get_hardware_profile() instead.
    pub fn get_profile() -> HashMap {
        // Legacy method - use get_hardware_profile() instead.
        let mut profile = get_hardware_profile(/* detailed= */ false);
        HashMap::from([("type".to_string(), profile["cpu".to_string()]["vendor".to_string()]), ("ram_gb".to_string(), profile["memory".to_string()]["total_gb".to_string()]), ("vram_mb".to_string(), if profile["gpu".to_string()] { profile["gpu".to_string()]["vram_mb".to_string()] } else { 0 }), ("threads".to_string(), profile["cpu".to_string()]["logical_cores".to_string()]), ("gpu".to_string(), if profile["gpu".to_string()] { profile["gpu".to_string()]["name".to_string()] } else { None })])
    }
}

/// Detect CPU vendor, cores, frequency.
/// 
/// Returns:
/// Dict with keys: vendor, name, logical_cores, physical_cores, freq_mhz
pub fn detect_cpu() -> Result<HashMap<String, Box<dyn std::any::Any>>> {
    // Detect CPU vendor, cores, frequency.
    // 
    // Returns:
    // Dict with keys: vendor, name, logical_cores, physical_cores, freq_mhz
    let mut cpu_info = HashMap::from([("vendor".to_string(), "Unknown".to_string()), ("name".to_string(), (platform.processor() || "Unknown".to_string())), ("logical_cores".to_string(), (os::cpu_count() || 1)), ("physical_cores".to_string(), (os::cpu_count() || 1)), ("freq_mhz".to_string(), None)]);
    // try:
    {
        // TODO: import psutil
        let mut physical = psutil.cpu_count(/* logical= */ false);
        if physical {
            cpu_info["physical_cores".to_string()] = physical;
        }
        let mut freq = psutil.cpu_freq();
        if freq {
            cpu_info["freq_mhz".to_string()] = ((freq.current as f64) * 10f64.powi(0)).round() / 10f64.powi(0);
        }
    }
    // except ImportError as exc:
    let mut name_upper = cpu_info["name".to_string()].to_uppercase();
    if name_upper.contains(&"INTEL".to_string()) {
        cpu_info["vendor".to_string()] = "Intel".to_string();
    } else if name_upper.contains(&"AMD".to_string()) {
        cpu_info["vendor".to_string()] = "AMD".to_string();
    } else if name_upper.contains(&"ARM".to_string()) {
        cpu_info["vendor".to_string()] = "ARM".to_string();
    }
    Ok(cpu_info)
}

/// Detect RAM availability and usage.
/// 
/// Returns:
/// Dict with keys: total_gb, available_gb, used_percent
pub fn detect_memory() -> Result<HashMap<String, Box<dyn std::any::Any>>> {
    // Detect RAM availability and usage.
    // 
    // Returns:
    // Dict with keys: total_gb, available_gb, used_percent
    let mut mem_info = HashMap::from([("total_gb".to_string(), 8.0_f64), ("available_gb".to_string(), 4.0_f64), ("used_percent".to_string(), 50.0_f64)]);
    // try:
    {
        // TODO: import psutil
        let mut vm = psutil.virtual_memory();
        mem_info["total_gb".to_string()] = (((vm.total / (1024).pow(3 as u32)) as f64) * 10f64.powi(1)).round() / 10f64.powi(1);
        mem_info["available_gb".to_string()] = (((vm.available / (1024).pow(3 as u32)) as f64) * 10f64.powi(1)).round() / 10f64.powi(1);
        mem_info["used_percent".to_string()] = vm.percent;
    }
    // except ImportError as _e:
    Ok(mem_info)
}

/// Detect dedicated GPU (Windows only with PowerShell/WMI).
/// 
/// Returns:
/// Dict with keys: vendor, name, vram_mb
/// Or None if no discrete GPU found
pub fn detect_gpu() -> Result<Option<HashMap<String, Box<dyn std::any::Any>>>> {
    // Detect dedicated GPU (Windows only with PowerShell/WMI).
    // 
    // Returns:
    // Dict with keys: vendor, name, vram_mb
    // Or None if no discrete GPU found
    if sys::platform != "win32".to_string() {
        None
    }
    // try:
    {
        let mut cmd = vec!["powershell".to_string(), "-NoProfile".to_string(), "-Command".to_string(), "Get-CimInstance Win32_VideoController | Where-Object {$_.Name -notmatch \"Remote|RDP\"} | Select-Object Name, AdapterRAM | ConvertTo-Json".to_string()];
        let mut output = subprocess::check_output(cmd, /* shell= */ false, /* text= */ true, /* stderr= */ subprocess::DEVNULL, /* timeout= */ 5).trim().to_string();
        if !output {
            None
        }
        let mut gpus = serde_json::from_str(&output).unwrap();
        if !/* /* isinstance(gpus, list) */ */ true {
            let mut gpus = vec![gpus];
        }
        for gpu_raw in gpus.iter() {
            let mut name = gpu_raw.get(&"Name".to_string()).cloned().unwrap_or("".to_string()).to_uppercase();
            if vec!["INTEL".to_string(), "BASIC".to_string(), "UHD".to_string(), "IRIS".to_string()].iter().map(|x| name.contains(&x)).collect::<Vec<_>>().iter().any(|v| *v) {
                continue;
            }
            let mut vram_bytes = gpu_raw.get(&"AdapterRAM".to_string()).cloned().unwrap_or(0);
            let mut vram_mb = 1.max((vram_bytes / (1024).pow(2 as u32)));
            let mut gpu_info = HashMap::from([("name".to_string(), gpu_raw.get(&"Name".to_string()).cloned().unwrap_or("Unknown".to_string())), ("vram_mb".to_string(), vram_mb), ("vendor".to_string(), "Unknown".to_string())]);
            if (name.contains(&"NVIDIA".to_string()) || name.contains(&"GEFORCE".to_string()) || name.contains(&"TESLA".to_string())) {
                gpu_info["vendor".to_string()] = "NVIDIA".to_string();
            } else if (name.contains(&"AMD".to_string()) || name.contains(&"RADEON".to_string()) || name.contains(&"EPYC".to_string())) {
                gpu_info["vendor".to_string()] = "AMD".to_string();
            } else if name.contains(&"INTEL".to_string()) {
                gpu_info["vendor".to_string()] = "Intel".to_string();
            }
            gpu_info
        }
    }
    // except Exception as e:
}

/// Recommend LLM tier based on hardware.
/// 
/// Args:
/// ram_gb: Total system RAM in GB
/// cpu_cores: Logical CPU cores
/// 
/// Returns:
/// Dict with: tier, max_model_params, quantization, n_gpu_layers, note
pub fn get_recommendation_tier(ram_gb: f64, cpu_cores: i64) -> HashMap<String, Box<dyn std::any::Any>> {
    // Recommend LLM tier based on hardware.
    // 
    // Args:
    // ram_gb: Total system RAM in GB
    // cpu_cores: Logical CPU cores
    // 
    // Returns:
    // Dict with: tier, max_model_params, quantization, n_gpu_layers, note
    if ram_gb < 4 {
        let mut tier = "potato".to_string();
        let mut max_params = "1B-3B".to_string();
        let mut quant = "Q4_K_S".to_string();
        let mut gpu_layers = 0;
        let mut note = "Very tight resources. Consider 1B models like TinyLlama".to_string();
    } else if ram_gb < 8 {
        let mut tier = "budget".to_string();
        let mut max_params = "3B-7B".to_string();
        let mut quant = "Q4_K_S or Q5_K_M".to_string();
        let mut gpu_layers = 0;
        let mut note = "Focus on 3B-7B quantized models".to_string();
    } else if ram_gb < 16 {
        let mut tier = "mid_range".to_string();
        let mut max_params = "7B-13B".to_string();
        let mut quant = "Q5_K_M or Q6_K".to_string();
        let mut gpu_layers = if ram_gb > 12 { 10 } else { 5 };
        let mut note = "Good range: 7B Mistral, Llama 2-7B".to_string();
    } else if ram_gb < 32 {
        let mut tier = "comfortable".to_string();
        let mut max_params = "13B-34B".to_string();
        let mut quant = "Q5_K_M, Q6_K, or Q8_0".to_string();
        let mut gpu_layers = if ram_gb > 24 { 20 } else { 15 };
        let mut note = "Can run larger models: Llama 2-13B, Mistral 7B without Q".to_string();
    } else {
        let mut tier = "rich".to_string();
        let mut max_params = "34B-70B+".to_string();
        let mut quant = "Q6_K, Q8_0, or full precision".to_string();
        let mut gpu_layers = 40;
        let mut note = "Plenty of resources: Llama 2-70B, Code Llama, Mixtral".to_string();
    }
    HashMap::from([("tier".to_string(), tier), ("ram_gb".to_string(), ram_gb), ("cpu_cores".to_string(), cpu_cores), ("max_model_params".to_string(), max_params), ("recommended_quant".to_string(), quant), ("n_gpu_layers".to_string(), gpu_layers), ("note".to_string(), note)])
}

/// Get complete hardware profile for ZEN_RAG.
/// 
/// Args:
/// detailed: If true, include recommendation tier and extra info
/// 
/// Returns:
/// Comprehensive hardware dictionary:
/// {
/// "platform": "Windows/Linux/Darwin",
/// "architecture": "x86_64/arm64/etc",
/// "cpu": {...},
/// "memory": {...},
/// "gpu": {...} or None,
/// "recommendation": {...} if detailed=true
/// }
pub fn get_hardware_profile(detailed: bool) -> HashMap<String, Box<dyn std::any::Any>> {
    // Get complete hardware profile for ZEN_RAG.
    // 
    // Args:
    // detailed: If true, include recommendation tier and extra info
    // 
    // Returns:
    // Comprehensive hardware dictionary:
    // {
    // "platform": "Windows/Linux/Darwin",
    // "architecture": "x86_64/arm64/etc",
    // "cpu": {...},
    // "memory": {...},
    // "gpu": {...} or None,
    // "recommendation": {...} if detailed=true
    // }
    let mut profile = HashMap::from([("platform".to_string(), sys::platform), ("architecture".to_string(), platform.machine()), ("cpu".to_string(), detect_cpu()), ("memory".to_string(), detect_memory()), ("gpu".to_string(), detect_gpu())]);
    if detailed {
        profile["recommendation".to_string()] = get_recommendation_tier(profile["memory".to_string()]["total_gb".to_string()], profile["cpu".to_string()]["logical_cores".to_string()]);
    }
    profile
}

/// Quick hardware summary for CLI tools (cached).
/// 
/// Used by: api_server /health endpoint
/// Returns minimal key info for status checks.
pub fn get_hardware_summary() -> HashMap<String, Box<dyn std::any::Any>> {
    // Quick hardware summary for CLI tools (cached).
    // 
    // Used by: api_server /health endpoint
    // Returns minimal key info for status checks.
    let mut profile = get_hardware_profile();
    let mut summary = HashMap::from([("ram_total_gb".to_string(), profile["memory".to_string()]["total_gb".to_string()]), ("ram_available_gb".to_string(), profile["memory".to_string()]["available_gb".to_string()]), ("cpu_count".to_string(), profile["cpu".to_string()]["logical_cores".to_string()]), ("cpu_physical".to_string(), profile["cpu".to_string()]["physical_cores".to_string()]), ("cpu_vendor".to_string(), profile["cpu".to_string()]["vendor".to_string()])]);
    if profile["gpu".to_string()] {
        summary["gpu_name".to_string()] = profile["gpu".to_string()]["name".to_string()];
        summary["gpu_vram_mb".to_string()] = profile["gpu".to_string()]["vram_mb".to_string()];
    }
    summary
}
